### PR TITLE
GitHub app authentication

### PR DESCRIPTION
**Pull request by ebarriosjr**
## Hey, I just made a Pull Request!

In this PR I added a way to authenticate backstage via a githubApp.

#### :heavy_check_mark: Checklist

- [ ] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

### CODE REVIEW

File: plugins/catalog-backend/src/ingestion/HigherOrderOperations.ts [hunk 1]
```diff
@@ -141,13 +142,40 @@ export class HigherOrderOperations implements HigherOrderOperation {
     );
   }
 
+  // Gets an application token to refresh github locations
+  private async getToken(): Promise<string> {
+    if (process.env.AUTH_GITHUB_TOKEN) {
+      return process.env.AUTH_GITHUB_TOKEN;
+    }
```
Comment (OrkoHunter) [inline plugins/catalog-backend/src/ingestion/HigherOrderOperations.ts:149]: Can we use the `GITHUB_TOKEN` here that we already have in place?
https://github.com/backstage/backstage/blob/a2b83df986c47d77aca65e9c5aa951e48b558e29/app-config.yaml#L81-L85

**Comment by OrkoHunter** [comment]
Hi @ebarriosjr, I'm looking at https://backstage.io/docs/tutorials/quickstart-app-auth and there seems to be a way how we can setup authentication using a GitHub app.

I am not able to understand what is the different approach here in this PR, if any. Excuse me about that. Could you maybe write some docs about how one can use this? Looks like you are adding some new configs like `APP_ID`, `PRIVATE_KEY`, `INSTALLATION_ID`, etc. This could really benefit some documentation/tutorial. Thank you!

**Comment by ebarriosjr** [comment]
Hey @OrkoHunter, 
I am not aware of any way to use a githubApp right now. These is still a work in progress. I just push it as PR to get some feedback out of the community.
I am working on tests and documentation right now. :)

File: plugins/catalog-backend/src/ingestion/HigherOrderOperations.ts (hunk 1 above)
Comment (ebarriosjr) [reply plugins/catalog-backend/src/ingestion/HigherOrderOperations.ts]: Replace it with GITHUB_TOKEN...
Thanks!

**Comment by benjdlambert** [comment]
Hey @ebarriosj! Thanks for this! Been excited to take a look through this for some time. Wanna set aside some time tomorrow to look through this as I've signed off for the day today. Will get back to you tomorrow!

**Comment by ebarriosjr** [comment]
Hi @benjdlambert , did you by any chance have time to review this?

**Comment by benjdlambert** [comment]
Sorry yes - I think that we were planning on trying to avoid passing the token everywhere. Ideally reader should just call out to get a token from a wrapper around the App Authentication. Is that what we said @Rugvip?

**Comment by benjdlambert** [comment]
I've also been looking at how we can make it a little less config driven, so that you can install the github app to your organisation or repo simply by using backstage rather than creating the app separately. But that's something we can address in the future.

**Comment by freben** [comment]
Hi! We went the rounds through the issues and PRs today and had a look here again.

It seems to me that there is some architecture and questions to be settled on our part, and I realise that this puts you in an uncomfortable situation. At this point, I would like to suggest that we put this effort a bit on hold. We are thankful for the effort you put into this, and it is something that we can catapult off of when we get the time to zone in on the github app auth situation. Does that sound OK?

**Comment by ebarriosjr** [comment]
Hi @freben 
it is ok, I have implemented this for our repo and would love to get something that works with githubApp on the master branch. But I do get that maybe this is not the most elegant way to go. Should I keep this branch up to date with master or just let it die?

**Comment by freben** [comment]
If it's OK with you, you can close. I wouldn't want to put it on anyone to keep rebasing, and github keeps the commits alive so we can use cherry picking or copy-paste when the time comes.

**Comment by ebarriosjr** [comment]
Ok, closing. I added some test that maybe could be copied to master if you guys want to FYI @freben

**Comment by freben** [comment]
🙇‍♂️

**Pull request closed without merging**